sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExecBase.scala

```diff
@@ -78,96 +77,265 @@ abstract class GpuWindowInPandasExecMetaBase(
-            table
-              .groupBy(partitionIndices:_*)
-              .aggregate(Aggregation.count(true).onColumn(0))
-          }
+      val groupLens = genGroupInfoFromBatch(batch)
+      if (isSplitGroups && groupLens.size > 1) {
+        // Split the groups when being required and there are more than one group.
+        val groupBatchesAndSizes = splitBatchPerGroups(batch, groupLens)
+          .zip(groupLens.map(Seq(_)))
+        val tails = groupBatchesAndSizes.tail.map {
+          case (sb, gpInfo) =>
+            (SpillableColumnarBatch(sb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY, spillCallback),
+              gpInfo)
         }
-        val orderedTable = withResource(cntTable) { table =>
-          table.orderBy(partitionIndices.map(id => Table.asc(id, true)): _*)
+        // Return the first one and enqueue others
+        queueBatches.enqueue(tails: _*)
+        groupBatchesAndSizes.head
+      } else {
+        (batch, groupLens)
+      }
+    }
+    (nextBatch, buildInfoBatchForPython(nextBatch, gpLens))
+  }
+
+  private[this] def splitBatchPerGroups(
+      batch: ColumnarBatch,
+      gpLens: Seq[Int]): Seq[ColumnarBatch] = {
+    val splitIndices = gpLens.dropRight(1).scan(0)(_+_).tail
+    // Close the original big batch
+    withResource(batch) { cb =>
+      withResource(GpuColumnVector.from(cb)) { table =>
+        withResource(table.contiguousSplit(splitIndices: _*)) { tables =>
+          val dataTypes = GpuColumnVector.extractTypes(batch)
+          tables.safeMap(t => GpuColumnVectorFromBuffer.from(t, dataTypes))
         }
-        val (countHostCol, numRows) = withResource(orderedTable) { table =>
-          // Yes copying the data to host, it would be OK since just copying the aggregated
-          // count column.
-          (table.getColumn(table.getNumberOfColumns - 1).copyToHost(), table.getRowCount)
+      }
+    }
+  }
+
+  private[this] def buildInfoBatchForPython(
+      batch: ColumnarBatch,
+      gpLens: Seq[_]): ColumnarBatch = {
+
+    // The rolling size may need tuning.
+    val rollingSize = 20000
+    val gpInfo: Seq[_] = if (gpLens.length <= rollingSize) {
+      gpLens
+    } else {
+      // Two levels concatenation to avoid too many small column vectors on GPU and
+      // exceeding GC overhead limit.
+      gpLens.sliding(rollingSize, rollingSize).toSeq
+    }
+
+    // Build the window bounds for bounded frames
+    val boundsCVs = windowFrames.filterNot(_.isUnbounded).flatMap { frame =>
+      frame.frameType match {
+        case RowFrame =>
+          val (lowerCV, upperCV, _) = genColumnsForWindowBounds(frame, gpInfo, 0)
+          Seq(lowerCV, upperCV)
+        case f => throw new UnsupportedOperationException(s"Unsupported window frame $f")
+      }
+    }.map(GpuColumnVector.from(_, IntegerType))
+    new ColumnarBatch(boundsCVs.toArray, batch.numRows())
+  }
+
+  /*
+   * (Currently do it be leveraging the existing APIs in cuDF Java.
```

Comment:
I would prefer to have an actual explanation of what this API is supposed to do.  This code is rather complex now and the explanation I think would be more helpful.